csharp/src/Ice/OutgoingRequestFrame.cs

```diff
@@ -100,23 +100,70 @@ public sealed class OutgoingRequestFrame : OutputStream
+            Size = Ice1Definitions.HeaderSize + 4 + Data.GetByteCount();
+            PayloadSize = OutputStream.Distance(Data, PayloadStart, payloadEnd);
+            PayloadEnd = payloadEnd;
+            IsSealed = true;
+        }
+
+        /// <summary>Creates and returns an OutputStream that can be used to write the payload of this request,
+        /// once the caller finish writting the payload it must call SavePayload passing the returned
+        /// stream as argument.</summary>
+        /// <param name="format">The format type for the payload.</param>
+        /// <returns>An OutputStream instance that can be used to write the payload of this request.</returns>
+        public override OutputStream WritePayload(FormatType? format = null)
+        {
+            if (Ostr != null)
+            {
+                throw new InvalidOperationException("the frame already start writting the payload");
             }
+
+            if (PayloadEnd != null)
+            {
+                throw new InvalidOperationException("the frame already contains a paylod");
+            }
+            Ostr = new OutputStream(Encoding, Data, PayloadStart);
+            Ostr.StartEncapsulation(Encoding, format);
+            return Ostr;
         }
 
-        /// <summary>Starts writing the parameters for this request.</summary>
-        /// <param name="format">The format for the parameters, null (meaning keep communicator's setting), SlicedFormat
-        /// or CompactFormat.</param>
-        public void StartParameters(FormatType? format = null)
-            => StartEncapsulation(_payloadEncoding, format);
+        private ArraySegment<byte> CreateRequestHeader(int requestId)
+        {
+            Debug.Assert(IsSealed);
+            ArraySegment<byte> requestHeader = new byte[Ice1Definitions.HeaderSize + 4];
+            int pos = 0;
+            Ice1Definitions.Magic.CopyTo(requestHeader.AsSpan(pos, Ice1Definitions.Magic.Length));
+            pos += Ice1Definitions.Magic.Length;
+            Ice1Definitions.ProtocolBytes.CopyTo(requestHeader.AsSpan(pos, Ice1Definitions.ProtocolBytes.Length));
+            pos += Ice1Definitions.ProtocolBytes.Length;
+            requestHeader[pos++] = (byte)Ice1Definitions.MessageType.RequestMessage;
+            pos++; // compression plabceholder
+            OutputStream.WriteInt(Size, requestHeader.AsSpan(pos, 4));
+            pos += 4;
+            OutputStream.WriteInt(requestId, requestHeader.AsSpan(pos, 4));
+
+            return requestHeader;
+        }
 
-        /// <summary>Marks the end of the parameters.</summary>
-        public void EndParameters() => EndEncapsulation();
+        internal List<ArraySegment<byte>> GetRequestData(int requestId)
+        {
+            var data = new List<ArraySegment<byte>>(Data.Count + 1)
+            {
+                CreateRequestHeader(requestId)
+            };
+            data.AddRange(Data);
+            return data;
```

Comment:
`GetRequestData` does not belong to this class, so the question is moot. If we kept it there, the array should not be a shared array that we write and rewrite each time we resent the request as it would make this OutgoingRequestFrame not readonly once sealed.

Reply:
Replaced by static method `Ice1Definitions.GetRequestData` still allocate the request header with each call, can the request handler pass a pre-allocate header and reuse it?